### PR TITLE
librbd: use custom allocator for aligned boost::lockfree::queue

### PR DESCRIPTION
**Pull request by dillaman**
If tcmalloc is in-use as the allocator and its version is less than 2.6.2,
it might be missing support for 'aligned_alloc'. This can result in the
glibc version of 'aligned_alloc' being used to allocate memory that is
then freed by tcmalloc -- resulting in a crash.

Fixes: http://tracker.ceph.com/issues/39703
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/allocator.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  template <class U>
+  allocator(const allocator<U>& other) noexcept : std::allocator<T>(other) {
+  }
+
+  pointer allocate(size_type n, const void* hint = nullptr) {
+    if (n > this->max_size()) {
+      throw std::bad_alloc();
+    }
+
+    if (alignof(T) > 1) {
```
Comment (tchaikov) [inline src/common/allocator.h]: could use
```c++
if constexpr alignof(T) > __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
// ...
}
```

Comment (dillaman) [reply src/common/allocator.h]: I didn’t want to use a potentially private macro. Is that actually part of the standard? The “__” makes me think it isn’t.

Comment (tchaikov) [reply src/common/allocator.h]: yeah, it is. see http://eel.is/c++draft/cpp.predefined#1.6 and https://en.cppreference.com/w/cpp/preprocessor/replace . it was added by http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0035r4.html .

Comment (dillaman) [reply src/common/allocator.h]: updated

Comment (tchaikov) [reply src/common/allocator.h]: could use `constexpr if` to enforce the compile-time condition.

Comment (dillaman) [reply src/common/allocator.h]: GCC 8 and 9 were both smart enough to not need it to optimize out the code.

Comment (tchaikov) [reply src/common/allocator.h]: fair enough.

**Pull request merged**